org.eclipse.xtext.xtext.ui.examples/projects/fowlerdsl/org.eclipse.xtext.example.fowlerdsl.tests/src/org/eclipse/xtext/example/fowlerdsl/tests/StatemachineFormatterTest.java

```diff
@@ -0,0 +1,317 @@
+/**
+ * Copyright (c) 2019, 2021 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.example.fowlerdsl.tests;
+
+import org.eclipse.xtext.testing.InjectWith;
+import org.eclipse.xtext.testing.XtextRunner;
+import org.eclipse.xtext.testing.formatter.FormatterTestHelper;
+import org.eclipse.xtext.testing.formatter.FormatterTestRequest;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import com.google.inject.Inject;
+
+/**
+ * @author miklossy - Initial contribution and API
+ */
+@RunWith(XtextRunner.class)
+@InjectWith(StatemachineInjectorProvider.class)
+public class StatemachineFormatterTest {
+	@Inject
+	private FormatterTestHelper formatterTestHelper;
+	
+	private static final String NL = System.lineSeparator();
```

Comment:
i am not sure if i like this.
we could also
- call toUnitLineSeparator on all cases
- uses internal MapBasedPreference to configure newLine
- have a FormatterTestHelper that does does it internal

@szarnekow @miklossy what do you think?

Reply:
I prefer option (3)

Reply:
Another one would be to use format strings with `%n` but it's kind of nasty since it still needs string concat.

Reply:
created https://github.com/eclipse/xtext-core/issues/1703 for the map based preference values api problem

Reply:
maybe something like

```
public class FormatterTestHelper {

	public static class Platform extends FormatterTestHelper {
		@Override
		public void assertFormatted(FormatterTestRequest req) {
			req.setToBeFormatted(Strings.toPlatformLineSeparator(req.getToBeFormatted()));
			CharSequence expectation = req.getExpectation();
			if (expectation != null) {
				req.setExpectation(Strings.toPlatformLineSeparator(expectation));
			}
			super.assertFormatted(req);
		}
	}
```

Reply:
Makes sense to me. Is the rewrite of the request itself necessary?

Reply:
the problem is that the original api is very bad to intercept as it is
public void assertFormatted(Procedure1<FormatterTestRequest> init) {
maybe you have a better idea?

Reply:
IMHO: it is crucial that such tests are easy to write and maintain. If you use the JUnit view failure of an assertEquals with strings, you can copy the actual value (with newlines) and paste it in a Java editor and Eclipse will add the corresponding string concatenation and the `\n` character at the end. If I don't use Xtend I keep the concatenated string generated by Eclipse in the editor and have a utility function to deal with possible different line endings in different operating systems. Using a solution with explicit NL makes these tests hard to write and hard to maintain...

Reply:
yes. an approach using the preferences (when made api) could be

```
	@Inject
	private FormatterTestHelper formatterTestHelper;
	
	public void assertFormatted(Procedure1<FormatterTestRequest> init) {
		formatterTestHelper.assertFormatted(new Procedure1<FormatterTestRequest>() {

			@Override
			public void apply(FormatterTestRequest p) {
				init.apply(p);
				p.preferences(m -> m.put(FormatterPreferenceKeys.lineSeparator, "\n"));
				
			}
		});
	}
	```
	
	then  it should work with "\n" as copy&pastable too

Reply:
this change also would be test local

Reply:
Looks way better to me.